nixos/modules/services/continuous-integration/buildkite-agents.nix

```diff
@@ -44,7 +45,7 @@ in
       };
 
       dataDir = mkOption {
-        default = "/var/lib/buildkite-agent";
+        default = "/var/lib/buildkite-agent-${name}";
         description = "The workdir for the agent";
         type = types.str;
       };
```

Comment:
do you know usecases of where this should be somewhere else than at this location?

If we restrict it to `/var/lib/buildkite-agent-${name}`, we could just set `StateDirectory=buildkite-agent=${name}`, and `DynamicUser=true`, and don't need to setup a system-wide user at all.

Reply:
Yes, we use the buildkite directory to store some gc roots that are used for Continuous Delivery. We could change to a ReadWriteDirectory, but I don't think `DynamicUser` is really set up for communication with the outside world.

Reply:
Our approach to CD is arguably horrible by abusing this.

Reply:
Well, I'd recommend using some post-build-hook on the global nix-daemon to sign and upload to a binary cache, but hmm… I can understand you can't change this that quick.

Still, what about defaulting to use `DynamicUser=true`, and `StateDirectory=buidlkite-agent-${name}`, and only use a system-wide user if both a `user` attribute and that `dataDir` attribute is set?

Reply:
> using some post-build-hook on the global nix-daemon to sign and upload to a binary cache

Been there, tried that :(. Builds become horribly slow and the host becomes almost unusable as it is uploading _everything_, even your user environment after `nix-env -i` (and that’s slow too).

Reply:
Well, the naive approach blocks the build loop, as documented in https://nixos.org/nix/manual/#chap-post-build-hook-caveats .

In your case, you might want to have the executed hook to fork away as fast as possible.

However, I see that discussion as not really related to this PR - it doesn't yet introduce `DynamicUser`.